### PR TITLE
8379327: 128-bit multiplication uses two multiply instructions on x86_64

### PR DESCRIPTION
**Pull request by IntegralPilot**
Introduces `MulHiLoL` and `UMulHiLoL` nodes to fuse 64-bit high and low multiply operations. During `final_graph_reshaping`, C2 now detects matching `MulHiL/UMulHiL` and `MulL` pairs and subsumes them into these new multi-nodes. The x86_64 backend is updated to emit a single `imulq` or `mulq` instruction for the result, with new tests verifying both correctness and code generation. I've verified the test work on both an x86 (testing it passes) and arm (testing it skips) device.

- [x] I confirm that I make this contribution in accordance with the [OpenJDK Interim AI Policy](https://openjdk.org/legal/ai).

---------
### Progress
- \[x\] Change must not contain extraneous whitespace
- \[x\] Commit message must refer to an issue
- \[x\] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8379327](https://bugs.openjdk.org/browse/JDK-8379327): 128-bit multiplication uses two multiply instructions on x86_64 (**Enhancement** - P4)

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/30174/head:pull/30174` \
`$ git checkout pull/30174`

Update a local copy of the PR: \
`$ git checkout pull/30174` \
`$ git pull https://git.openjdk.org/jdk.git pull/30174/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 30174`

View PR using the GUI difftool: \
`$ git pr show -t 30174`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/30174.diff">https://git.openjdk.org/jdk/pull/30174.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/30174#issuecomment-4035130532)
</details>

### CODE REVIEW

File: src/hotspot/share/opto/compile.cpp [hunk 1]
```diff
@@ -3259,6 +3259,41 @@ void Compile::handle_div_mod_op(Node* n, BasicType bt, bool is_unsigned) {
...
+    Node* rhs = n->in(2);
+    if (rhs != nullptr && rhs->outcnt() >= 2) {
+      for (DUIterator_Fast dmax, i = rhs->fast_outs(dmax); i < dmax; i++) {
+        Node* use = rhs->fast_out(i);
+        if (use != n &&
+            use->Opcode() == Op_MulL &&
+            use->req() == n->req() &&
+            use->in(1) == rhs &&
+            use->in(2) == lhs) {
+          mul = use;
```
Comment (dean-long) [inline src/hotspot/share/opto/compile.cpp]: This looks like a version of find_similar() that uses in(2) instead of in(1), but it also relaxes the requirement to match all inputs.  What if the control input is different?
Considering refactoring find_similar() so that it works with either in(1) or in(2).

File: src/hotspot/cpu/x86/x86.ad [hunk 2]
```diff
@@ -11647,6 +11647,58 @@ instruct mulL_mem_imm(rRegL dst, memory src, immL32 imm, rFlagsReg cr)
...
+%{
+  match(UMulHiLoL rax src);
+  effect(KILL cr);
+
+  ins_cost(300);
+  format %{ "mulq   RDX:RAX, RAX, $src\t# umulhilo" %}
+  ins_encode %{
+    __ mulq($src$$Register);
+  %}
+  ins_pipe(ialu_reg_reg_alu0);
```
Comment (dean-long) [inline src/hotspot/cpu/x86/x86.ad:11400]: What if src is RAX or RDX, then this instruction will kill it, right?  I assume that is why the DivMod rule uses no_rax_rdx_RegI.  Also, I don't see how theses rules specifies that the result will be in RDX:RAX, and further how which output register is associate with the hi value and which with the lo value.

File: src/hotspot/cpu/x86/x86.ad [hunk 3]
```diff
@@ -11647,6 +11647,58 @@ instruct mulL_mem_imm(rRegL dst, memory src, immL32 imm, rFlagsReg cr)
...
+  format %{ "imulq   RDX:RAX, RAX, $src\t# mulhilo" %}
+  ins_encode %{
+    __ imulq($src$$Register);
+  %}
+  ins_pipe(ialu_reg_reg_alu0);
+%}
+
+instruct umulHiLoL_rReg(rax_RegL rax, rdx_RegL rdx, rRegL src, rFlagsReg cr)
+%{
+  match(UMulHiLoL src rax);
```
Comment (dean-long) [inline src/hotspot/cpu/x86/x86.ad:11391]:
```suggestion
  match(UMulHiLoL src rax);
  match(UMulHiLoL rax src);
```
The match line seems to be the only difference between the two, so they can be combined.

File: src/hotspot/share/opto/mulnode.hpp [hunk 4]
```diff
@@ -205,6 +207,52 @@ class UMulHiLNode : public Node {
...
+
+  static MulHiLoLNode* make(Node* mul_hi);
+
+  ProjNode* lo_proj() { return proj_out_or_null(lo_proj_num); }
+  ProjNode* hi_proj() { return proj_out_or_null(hi_proj_num); }
+
+private:
+  virtual bool depends_only_on_test() const { return false; }
+};
+
```
Comment (dean-long) [inline src/hotspot/share/opto/mulnode.hpp:224]: This class looks almost exactly like DivModNode.  Would it make sense to refactor the two into a common superclass?

File: src/hotspot/share/opto/compile.cpp (hunk 1 above)
Comment (IntegralPilot) [reply src/hotspot/share/opto/compile.cpp]: Done, thanks for the idea!

File: src/hotspot/cpu/x86/x86.ad (hunk 3 above)
Comment (IntegralPilot) [reply src/hotspot/cpu/x86/x86.ad:11391]: Applied, thanks.

File: src/hotspot/share/opto/mulnode.hpp (hunk 4 above)
Comment (IntegralPilot) [reply src/hotspot/share/opto/mulnode.hpp:224]: Yes, I think it would. I've done it now. Thanks for the idea!

File: src/hotspot/cpu/x86/x86.ad (hunk 2 above)
Comment (IntegralPilot) [reply src/hotspot/cpu/x86/x86.ad:11400]: > What if src is RAX or RDX, then this instruction will kill it, right? I assume that is why the DivMod rule uses no_rax_rdx_RegI

Oh yes, that's right, sorry. I've fixed it now.

> Also, I don't see how theses rules specifies that the result will be in RDX:RAX, and further how which output register is associate with the hi value and which with the lo value.

See `MulHiLoLNode::match` in `src/hotspot/share/opto/mulnode.cpp`

File: src/hotspot/share/opto/node.cpp [hunk 5]
```diff
@@ -2875,17 +2875,30 @@ bool Node::is_iteratively_computed() {
...
+                if (use->in(j) != in(j)) {
+                  same_inputs = false;
+                  break;
+                }
+              }
+              if (same_inputs) {
+                return use;
+              }
+            }
+          } else if (use->has_same_inputs_as(this)) {
```
Comment (dean-long) [inline src/hotspot/share/opto/node.cpp]: Actually, I don't think we need "swapped" or to look at in(2) at all.  We should be able to find that match through either input.  And I don't think we need to do the search twice.  We just need a version that takes the commutative property of multiply into account and compares in(1) and in(2) appropriately:
```
(use->in(1) == in(2) && use->in(2) == in(1)) || (use->in(1) == in(1) && use->in(2) == in(2))
```

Comment (IntegralPilot) [reply src/hotspot/share/opto/node.cpp]: Oh yes, good idea, thanks. Done now.

File: src/hotspot/cpu/x86/x86.ad (hunk 2 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:11400]: You actually do not need `no_rax_rdx_RegL` here. In fact, this would restrict us from doing `MulHiLo` with 2 operands being the same value. If `src` is assigned `rax` or `rdx`, then it is killed, the register allocator must know that, because `rax` and `rdx` are defined right below. The same thing happens for any other arithmetic instruction.

    instruct addI_rReg(rRegI dst, rRegI src, rFlagsReg cr)

If `src` and `dst` are the same, then `src` is killed. We do not need to specify that `dst` and `src` must be different. You only need to use `no_...` when the instruction cannot be executed with that register.

File: src/hotspot/cpu/x86/x86.ad [hunk 6]
```diff
@@ -11647,6 +11647,34 @@ instruct mulL_mem_imm(rRegL dst, memory src, immL32 imm, rFlagsReg cr)
...
+  format %{ "imulq   RDX:RAX, RAX, $src\t# mulhilo" %}
+  ins_encode %{
+    __ imulq($src$$Register);
+  %}
+  ins_pipe(ialu_reg_reg_alu0);
+%}
+
+instruct umulHiLoL_rReg(rax_RegL rax, rdx_RegL rdx, no_rax_rdx_RegL src, rFlagsReg cr)
+%{
+  match(UMulHiLoL src rax);
```
Comment (merykitty) [inline src/hotspot/cpu/x86/x86.ad:11391]: These 2 rules are exactly the same, so there is no use specifying 2 of them. The difference between them is the set of possible registers, which are only available AFTER matching, so it cannot affect the process of matching.

File: test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusionX64.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+public class TestMultiplyHighLowFusionX64 {
```
Comment (merykitty) [inline test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusionX64.java]: Please use the IR framework instead.

File: src/hotspot/cpu/x86/x86.ad (hunk 6 above)
Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:11391]: Because we kill `rax`, and I believe we try to commute the operands of `MulL` that likely dies here to the first operand, I think it is better to keep `match(rax src)` and remove the other one.

Comment (dean-long) [reply src/hotspot/cpu/x86/x86.ad:11391]: If we remove `match(op src rax)` and the rhs is already in RAX, it will take extra register shufffling/spills to match with `match(op src rax)`, right?  In other words, I think register masks do affect matching.

File: src/hotspot/share/opto/node.cpp [hunk 8]
```diff
@@ -2875,17 +2875,32 @@ bool Node::is_iteratively_computed() {
...
+            same = true;
+          } else if (is_commutative && req() >= 3) {
+            if (use->in(0) == in(0) && use->in(1) == in(2) && use->in(2) == in(1)) {
+              same = true;
+              for (uint j = 3; j < req(); j++) {
+                if (use->in(j) != in(j)) {
+                  same = false;
+                  break;
+                }
+              }
```
Comment (dean-long) [inline src/hotspot/share/opto/node.cpp:2907]: This is getting close.  However, it still has 2 * req() worse case behavior, instead of req() + 2, right?  I suggest using `has_same_inputs_as` only for !is_commutative, and adjust your is_commutative logic to handle both swapped and not-swapped cases.

File: src/hotspot/cpu/x86/x86.ad (hunk 2 above)
Comment (IntegralPilot) [reply src/hotspot/cpu/x86/x86.ad:11400]: Oh thanks, I've changed it now.

File: test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusionX64.java (hunk 7 above)
Comment (IntegralPilot) [reply test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusionX64.java]: Migrated it over now, thank you.

File: src/hotspot/share/opto/node.cpp (hunk 8 above)
Comment (IntegralPilot) [reply src/hotspot/share/opto/node.cpp:2907]: Oh great idea, thanks, I've done it now.

File: src/hotspot/cpu/x86/x86.ad (hunk 6 above)
Comment (IntegralPilot) [reply src/hotspot/cpu/x86/x86.ad:11391]: I have kept both match rules for now based on @dean-long's point about avoiding extra register shuffling/spills if the RHS is already in RAX. Let me know if either of you would prefer a different approach!

Comment (merykitty) [reply src/hotspot/cpu/x86/x86.ad:11391]: That doesn't matter, because the fact that rhs is in `rax` is only available at register allocation, long after the matching have finished. And both `rRegI` and `rax_RegI` match `rax_RegI`, so the matcher cannot use that info to choose a suitable rule.

Comment (dean-long) [reply src/hotspot/cpu/x86/x86.ad:11391]: OK, so we want to give the hint to the register allocator, not the matcher, that if either the lhs or rhs side is already in RAX, leave it (don't swap).  It sounds like you're saying these two match rules do not achieve that goal.  @IntegralPilot can you test that out and see if the extra match rule does reduce register shuffling?  If it doesn't help, then I would suggest to remove "rax" from the match rule, use src1 and src2 instead, and at codegen time, copy src1 or src2 into RAX if needed.

Comment (IntegralPilot) [reply src/hotspot/cpu/x86/x86.ad:11391]: @dean-long @merykitty I tested out the behavior of the match rules to check for register shuffling.

I tried it with both match rules, and then just the `match(op rax src)` rule. 

Both times it produced the exact same assembly without any redundant shuffling, just a single needed move into `RAX`:
```assembly
01a     movq    RAX, RSI          # spill
01d     imulq   RDX:RAX, RAX, RDX # mulhilo
```

Because the register allocator is already optimally handling the setup for `RAX` using the `rax_RegL` definition (and not doing any unnecessary swapping), I don't think we need to fall back to using generic `src1/src2` and doing manual moves in the `ins_encode` block. 

I've gone ahead and removed the redundant extra match rules for both `MulHiLoL` and `UMulHiLoL` to keep the `.ad` file clean. Let me know if this looks good to you! :)

**Comment by IntegralPilot** [comment]
Just pushed new commit addressing unneeded rule, as per my testing documented in reply to the above thread.

File: src/hotspot/cpu/x86/x86.ad (hunk 6 above)
Comment (dean-long) [reply src/hotspot/cpu/x86/x86.ad:11391]: So in your example, src registers were RSI and RDX, and one got copied to RAX.  In that case a spill is always necessary.  But that doesn't help decide about the case where the left source or right source is already in RAX.

Comment (dean-long) [reply src/hotspot/cpu/x86/x86.ad:11391]: Making the test a little more complicated:
```
 private static volatile long one = 1;
 public static long doMath(long a, long b) {
   long x = Long.divideUnsigned(a, one);
   long y = Long.divideUnsigned(b, one);
   long low = x * y;
   long high = Math.multiplyHigh(x, y);
   return low + high;
}
```
and without the extra rule:
```
087     movq    R10, RAX	# spill
08a     movq    RAX, R11	# spill
08d     imulq   RDX:RAX, RAX, R10	# mulhilo
```
vs with the extra rule:
```
087     imulq   RDX:RAX, RAX, R11	# mulhilo
```

Comment (IntegralPilot) [reply src/hotspot/cpu/x86/x86.ad:11391]: Oh, thank you for picking that up and writing a better test. I'll add the extra rule back now.

Comment (dean-long) [reply src/hotspot/cpu/x86/x86.ad:11391]: FYI, I also tried adding "MulHiLoL" to the commutative operator list in MatchNode::count_commutative_op(), but it did not reduce the spills for some reason.

File: src/hotspot/share/opto/mulnode.cpp [hunk 9]
```diff
@@ -23,6 +23,8 @@
  */
 
 #include "memory/allocation.inline.hpp"
+#include "opto/machnode.hpp"
+#include "opto/matcher.hpp"
```
Comment (dean-long) [inline src/hotspot/share/opto/mulnode.cpp]: #includes need to be sorted.  Make sure to run test/hotspot/jtreg/sources/TestIncludesAreSorted.java

Comment (IntegralPilot) [reply src/hotspot/share/opto/mulnode.cpp]: Oh thank you, just pushed the sorted version now (which passes the test)! :)

**Review by dean-long — COMMENTED** [review]
Looks good.  Testing passed.

**Review by dean-long — APPROVED** [review]

File: test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8379327
+ * @summary Verify correctness for combined low/high 64-bit multiplication patterns.
+ * @library /test/lib /
+ * @run driver compiler.c2.TestMultiplyHighLowFusion
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java]:
```suggestion
 * @run driver ${test.main.class}
```
Avoid copy mistakes in the future.

File: test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+      -1L,
+      Long.MIN_VALUE,
+      Long.MAX_VALUE,
+      0x00000000FFFFFFFFL,
+      0xFFFFFFFF00000000L,
+      0x123456789ABCDEFL,
+      0xFEDCBA9876543210L
+  };
+
+  private static final java.util.Random RANDOM = new java.util.Random(0x8379327L);
```
Comment (eme64) [inline test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java]: Do you know about `Generators.java`? It has special random distributions that make sure edge-case values (like the ones you list above) are more frequently chosen.

**Review by eme64 — CHANGES_REQUESTED** [review]
Some drive-by comments.

Also: can you please add the `Math.multiplyHigh` and `Math.unsignedMultiplyHigh` to the template framework?
It would give us some additional test coverage - it would be good to have it now that we intrinsify the methods.

See:
`compiler/lib/template_framework/library/Operations.java`

After line:
`// TODO: Math and other classes.`

We plan to add all other methods sometime soon as well: https://bugs.openjdk.org/browse/JDK-8369696
(feel free to work on that in a separate RFE if you are interested)

File: test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java (hunk 10 above)
Comment (IntegralPilot) [reply test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java]: Thanks, applied.

File: test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java (hunk 11 above)
Comment (IntegralPilot) [reply test/hotspot/jtreg/compiler/c2/TestMultiplyHighLowFusion.java]: Oh I didn't! That's really useful, thanks. I've modified the code to use it now.

**Comment by sviswa7** [comment]
@IntegralPilot Nice work! Could you please add a jmh test micro as well?

**Review by dean-long — APPROVED** [review]

File: src/hotspot/cpu/x86/x86.ad (hunk 6 above)
Comment (sviswa7) [reply src/hotspot/cpu/x86/x86.ad:11391]: I looked through the adlc, the commutative rules are only generated when at least one of the operand is a more complex operand (e.g. memory operand).

File: src/hotspot/share/opto/divnode.hpp [hunk 12]
```diff
@@ -241,36 +241,27 @@ class UModLNode : public DivModIntegerNode {
...
-class DivModNode : public MultiNode {
+class DivModNode : public BinaryMultiNode {
 protected:
-  DivModNode( Node *c, Node *dividend, Node *divisor );
+  DivModNode(Node* ctrl, Node* dividend, Node* divisor) : BinaryMultiNode(ctrl, dividend, divisor) {}
 public:
   enum {
     div_proj_num =  0,      // quotient
     mod_proj_num =  1       // remainder
   };
```
Comment (sviswa7) [inline src/hotspot/share/opto/divnode.hpp]: We should standardize these to first_proj_num and second_proj_num and move the enum to BinaryMultiNode.

File: src/hotspot/share/opto/divnode.hpp [hunk 13]
```diff
@@ -241,36 +241,27 @@ class UModLNode : public DivModIntegerNode {
...
-  virtual Node *Ideal(PhaseGVN *phase, bool can_reshape) { return nullptr; }
-  virtual const Type* Value(PhaseGVN* phase)  const { return bottom_type(); }
-  virtual uint hash() const { return Node::hash(); }
-  virtual bool is_CFG() const  { return false; }
-  virtual uint ideal_reg() const { return NotAMachineReg; }
 
   static DivModNode* make(Node* div_or_mod, BasicType bt, bool is_unsigned);
 
   ProjNode* div_proj() { return proj_out_or_null(div_proj_num); }
   ProjNode* mod_proj() { return proj_out_or_null(mod_proj_num); }
```
Comment (sviswa7) [inline src/hotspot/share/opto/divnode.hpp]: These could be standardized to first_proj() and second_proj() and moved to BinaryMultiNode.

File: src/hotspot/share/opto/mulnode.cpp [hunk 14]
```diff
@@ -606,6 +608,36 @@ const Type* UMulHiLNode::Value(PhaseGVN* phase) const {
...
+  [[maybe_unused]] Node* lo_proj = new ProjNode(umul_hi_lo, MulHiLoLNode::lo_proj_num);
+  [[maybe_unused]] Node* hi_proj = new ProjNode(umul_hi_lo, MulHiLoLNode::hi_proj_num);
+  return umul_hi_lo;
+}
+
+Node* MulHiLoLNode::match(const ProjNode* proj, const Matcher* match) {
+  uint ideal_reg = proj->ideal_reg();
+  RegMask rm;
+  if (proj->_con == lo_proj_num) {
+    rm.assignFrom(match->divL_proj_mask());
```
Comment (sviswa7) [inline src/hotspot/share/opto/mulnode.cpp]: divL_proj_mask/modL_proj_mask could be standardized to firstL_proj_mask/secondL_proj_mask or some better name than div/mod specific.

**Comment by IntegralPilot** [comment]
/open

Hi, I am so sorry, I got quite busy with some things IRL and forgot about this! I have set aside some time later this week to implement the current round of feedback.

File: src/hotspot/share/opto/divnode.hpp (hunk 12 above)
Comment (IntegralPilot) [reply src/hotspot/share/opto/divnode.hpp]: Done, thanks for the suggestion!

File: src/hotspot/share/opto/divnode.hpp (hunk 13 above)
Comment (IntegralPilot) [reply src/hotspot/share/opto/divnode.hpp]: Done, thanks for the suggestion

File: src/hotspot/share/opto/mulnode.cpp (hunk 14 above)
Comment (IntegralPilot) [reply src/hotspot/share/opto/mulnode.cpp]: Sounds good, I've done it now, thanks

**Comment by IntegralPilot** [comment]
@sviswa7 thanks for the feedback, I've just applied it now and added the JMH test as suggested. On Azure `Standard D2s v5 (2 vcpus, 8 GiB memory)` (I've plenty of free Azure Student credits and thought it was better to test on standardised environment), I got the following results (done before I rebased):
| Benchmark | Mode | Score (ns/op) | Error (ns/op) | Description |
| :--- | :---: | :---: | :---: | :--- |
| `signedLowOnly` | avgt | 128.988 | ± 1.296 | Signed 64-bit lower multiplication only |
| `signedHighOnly` | avgt | 1547.599 | ± 3.092 | Signed 64-bit upper multiplication only |
| **`signedLowPlusHigh` (Fused)** | **avgt** | **906.441** | **± 8.073** | **Fused signed 128-bit multiplication** |
| `unsignedHighOnly` | avgt | 1546.914 | ± 1.626 | Unsigned 64-bit upper multiplication only |
| **`unsignedLowPlusHigh` (Fused)** | **avgt** | **901.727** | **± 3.271** | **Fused unsigned 128-bit multiplication** |

* For signed its about a **45.9% reduction** in execution time (1676.587 ns/op vs 906.441 ns/op) to merge them with this vs doing them individually.
* For unsigned its about a **46.2% reduction** (1675.902 ns/op vs 901.727 ns/op).

@rose00 thanks for your architectural thoughts, that was a useful read. I agree that it is important this extensible. Thus to make it extensible and the concerns re endian/arithmetic assumptions, I refactored the design in the commit I just pushed. I think this new version aligns with your thoughts in Option 2 and is the best path forward.

@dean-long With the latest push for the changes to address sviswa/rose's feedback, I think it invalidated your approval. If/when you have time, I'd appreciate a re-review if possible. :)

**Comment by dean-long** [comment]
Could you merge with the latest jdk master?

**Review by dean-long — APPROVED** [review]

**Comment by IntegralPilot** [comment]
@dean-long pushed merge commit now, thanks for the review!

**Review by dean-long — APPROVED** [review]

**Review by sviswa7 — APPROVED** [review]
Looks good to me.

**Comment by IntegralPilot** [comment]
/integrate

Thanks both!
